### PR TITLE
handle ConnectionErrors from koji.ClientSession

### PR DESCRIPTION
**Pull request by mlangsdorf**
koji.ClientSession is supposed to retrying ConnectionErrors, but it doesn't
appear to do so correctly. Add a wrapper to koji.ClientSession that will catch
ConnectionErrors generated by methods and retry the method up to 10 times,
only raising if there is still a ConnectionError or if there is some other
error.

Signed-off-by: Mark Langsdorf <mlangsdo@redhat.com>

### CODE REVIEW

File: atomic_reactor/koji_util.py [hunk 1]
```diff
@@ -40,6 +42,33 @@ def callback(self, offset, totalsize, size, t1, t2):  # pylint: disable=W0613
...
+
+    def __getattr__(self, name):
+        session_attr = getattr(self._wrapped_session, name)
+        if callable(session_attr):
+            def call_with_catch(*a, **kw):
+                for retry in range(10):
+                    try:
+                        return session_attr(*a, **kw)
+                    except ConnectionError as exc:
+                        continue
```
Comment (lcarva) [inline atomic_reactor/koji_util.py]: Should we add some sort of back off logic here?

**Comment by twaugh** [comment]
Yes, there is a `retry()` method in atomic_reactor/core.py that has the sort of logic we want. I don't think we can use it directly but we could make a new one for use with koji.

**Comment by mlangsdorf** [comment]
Added the retry logic, fixed the weird exception error for Python 3.

File: tests/test_koji_util.py [hunk 2]
```diff
@@ -101,16 +105,57 @@ def test_create_simple_session(self):
...
+        ('krb_login', {}),
+    ])
+    @pytest.mark.parametrize(('error_type', 'should_recover', 'attempts'), [
+        (ConnectionError, True, HTTP_MAX_RETRIES - 1),
+        (ConnectionError, False, HTTP_MAX_RETRIES),
+        (KeyError, False, 1),
+    ])
+    def test_create_session_failures(self, tmpdir, session_type, session_args,
+                                     error_type, should_recover, attempts):
+        url = 'https://koji-hub-url.com'
```
Comment (twaugh) [inline tests/test_koji_util.py]: Let's always use `.example.com` for domains in test cases.

Comment (mlangsdorf) [reply tests/test_koji_util.py]: Fixed.

**Comment by mlangsdorf** [comment]
Relocated the sleep backoff so it actually executes, and removed the raise from koji_login since it was impossible to trigger since the underlying functions either return True or raise errors.

**Comment by rcerven** [comment]
apart from mine 1 comment looks good

File: tests/test_koji_util.py [hunk 3]
```diff
@@ -96,21 +100,64 @@ def test_koji_login_ssl(self, tmpdir, proxyuser, serverca):
...
+            session.should_receive('ssl_login').and_return(True)
+        else:
+            session.should_receive('krb_login').once().and_return(True)
 
         (flexmock(koji_util.koji).should_receive('ClientSession').with_args(
             url, opts={'krb_rdns': False}).and_return(session))
-        assert create_koji_session(url, {}) == session
+        assert create_koji_session(url, args)._wrapped_session == session
+
+    @pytest.mark.parametrize(('session_type', 'session_args'), [
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply tests/test_koji_util.py]: Fixed.

File: atomic_reactor/koji_util.py [hunk 4]
```diff
@@ -40,6 +43,38 @@ def callback(self, offset, totalsize, size, t1, t2):  # pylint: disable=W0613
...
+                retry_delay = HTTP_BACKOFF_FACTOR
+                last_exc = None
+                for retry in range(HTTP_MAX_RETRIES):
+                    try:
+                        return session_attr(*a, **kw)
+                    except ConnectionError as exc:
+                        time.sleep(retry_delay * (2 ** retry))
+                        last_exc = exc
+                        continue
+                    except Exception as exc:
```
Comment (sosiouxme) [inline atomic_reactor/koji_util.py]: Why catch this at all? Should be the same result without it.

File: tests/test_koji_util.py [hunk 5]
```diff
@@ -96,21 +100,64 @@ def test_koji_login_ssl(self, tmpdir, proxyuser, serverca):
...
+    ])
+    def test_create_authenticated_session(self, tmpdir, ssl_session):
+        url = 'https://example.com'
+        args = {}
 
-    def test_create_authenticated_session(self):
-        url = 'https://koji-hub-url.com'
         session = flexmock()
-        session.should_receive('krb_login').once().and_return(True)
+        if ssl_session:
```
Comment (sosiouxme) [inline tests/test_koji_util.py]: Seems like an unrelated change... is it just to improve code coverage?

File: tests/test_koji_util.py [hunk 6]
```diff
@@ -96,21 +100,64 @@ def test_koji_login_ssl(self, tmpdir, proxyuser, serverca):
...
+            session.should_receive('ssl_login').and_return(True)
+        else:
+            session.should_receive('krb_login').once().and_return(True)
 
         (flexmock(koji_util.koji).should_receive('ClientSession').with_args(
             url, opts={'krb_rdns': False}).and_return(session))
-        assert create_koji_session(url, {}) == session
+        assert create_koji_session(url, args)._wrapped_session == session
+
+    @pytest.mark.parametrize(('auth_type', 'auth_args'), [
```
Comment (sosiouxme) [inline tests/test_koji_util.py]: Covers the same branch in auth logic as test case above; I'd rather this just focused on the one behavior of new interest, the exception retry.

Comment (mlangsdorf) [reply tests/test_koji_util.py]: The session wrapper code needs to handle 3 cases: genuine exception such as a KeyError, ConnectionError that times out, and momentary ConnectionError that clears on a retry. And it needs to do it for ssl and krb logins.

The clears on a retry eventually ends up in the same place as a successful session, but I think there's merit in testing both. Both to get 100% coverage and to show how the code is supposed to work.

File: atomic_reactor/koji_util.py (hunk 4 above)
Comment (mlangsdorf) [reply atomic_reactor/koji_util.py]: Fixed, thanks.

File: atomic_reactor/koji_util.py [hunk 7]
```diff
@@ -88,9 +120,7 @@ def koji_login(session,
 
         result = session.krb_login(**kwargs)
 
-    if not result:
-        raise RuntimeError('Unable to perform Koji authentication')
-
+    # koji.ClientSession.*_login raise errors or return True, so do the same here.
```
Comment (lcarva) [inline atomic_reactor/koji_util.py]: Why this change?

I think it's pretty bad that koji's lib doesn't raise an error if auth fails. This was a way to make its usage a little bit nicer. With this change, all callers must be modified to verify that auth actually succeeded.

Comment (mlangsdorf) [reply atomic_reactor/koji_util.py]: Doh! I missed that *_login can return False under some circumstances. I'll fix the code and the test case.

Comment (mlangsdorf) [reply atomic_reactor/koji_util.py]: Fixed.

File: atomic_reactor/koji_util.py [hunk 8]
```diff
@@ -40,6 +43,35 @@ def callback(self, offset, totalsize, size, t1, t2):  # pylint: disable=W0613
...
+        session_attr = getattr(self._wrapped_session, name)
+        if callable(session_attr):
+            def call_with_catch(*a, **kw):
+                retry_delay = HTTP_BACKOFF_FACTOR
+                last_exc = None
+                for retry in range(HTTP_MAX_RETRIES):
+                    try:
+                        return session_attr(*a, **kw)
+                    except ConnectionError as exc:
+                        time.sleep(retry_delay * (2 ** retry))
```
Comment (twaugh) [inline atomic_reactor/koji_util.py]: This will go:
* call -> exception
* time.sleep() ...
* call -> exception
* time.sleep() ......
* call -> exception
* time.sleep() ..................
* propagate exception

In other words, we are sleeping *after* each attempt, even the last. Don't we want to sleep *before* each attempt other than the first?

**Comment by mlangsdorf** [comment]
added to release notes under # Bugfixes:
- koji connection errrors will be retried a few times instead of ending a build immediately
